gpt4all-chat/database.cpp

```diff
@@ -1046,57 +1093,74 @@ size_t Database::chunkStream(QTextStream &stream, int folder_id, int document_id
     // TODO: implement line_from/line_to
     constexpr int line_from = -1;
     constexpr int line_to = -1;
-    QList<QString> words;
+
     int chunks = 0;
     int addedWords = 0;
 
-    for (;;) {
-        QString word;
-        stream >> word;
-        if (stream.status() && !stream.atEnd())
-            return -1;
-        charCount += word.length();
-        if (!word.isEmpty())
-            words.append(word);
-        if (stream.status() || charCount + words.size() - 1 >= m_chunkSize) {
-            if (!words.isEmpty()) {
-                const QString chunk = words.join(" ");
-                QSqlQuery q(m_db);
-                int chunk_id = 0;
-                if (!addChunk(q,
-                    document_id,
-                    chunk,
-                    file,
-                    title,
-                    author,
-                    subject,
-                    keywords,
-                    page,
-                    line_from,
-                    line_to,
-                    words.size(),
-                    &chunk_id
-                )) {
-                    qWarning() << "ERROR: Could not insert chunk into db" << q.lastError();
-                }
+    QString buffer;
+    buffer.reserve(m_chunkSize * 2);
 
-                addedWords += words.size();
+    // FIXME: Consider using regex here instead for instance ' \r\n' seems to be used to indicate
+    // an intentional formatting based line break rather than a semantic line break in some PDFs.
+    const QList<QString> splitters = {"\n\n", "\r\n\r\n", "\r\n", "\n", " "};
```

Comment:
any reason we _shouldn't_ use regex?

Reply:
Well, can you think of a reason that a space before a "\r\n" should be treated as something other than a word boundary?

Basically, this FIXME proposes to treat:

* " \r\n" as a word boundary
* "\r\n" as a line boundary

I'm looking at a specific PDF file that seems to use the former as a word boundary, but how widespread is this? Is it a safe assumption to make?

Reply:
We could also treat a space before "\n" as a word boundary as well...